### PR TITLE
Cert check should check for hostname after resolving well-known

### PR DESCRIPTION
**Pull request by alexte**
https://matrix.org/docs/spec/server_server/r0.1.1.html says correctly that a delegated host for a domain must provide a valid cert for the delegated hostname and not the domein.
Sorry but I couldn't test this patch.

### CODE REVIEW

**Comment by alexte** [comment]
I tested the source now and found that two more changes where necessary for the .well-known handling
1) pulling the certificate should be done with the resolved host:port from well-known
2) cert should be checked with host from well-known without port
I solved the the cert check for me, I hope it doesn't break the check for others.

**Comment by richvdh** [comment]
I think this duplicates the work in https://github.com/matrix-org/gomatrixserverlib/pull/109.

**Comment by alexte** [comment]
The main part in this patch is the ServerHost should be checked instead of then ServerName (eg matrix.mail.at and not mail.at) in the isValidCert, and I stripped the port in the calling function and not in the isValidCertificate function,
isValidCertificate shouldn't care about port nummber IMHO

**Comment by richvdh** [comment]
I don't think this PR is relevant any more

File: src/github.com/matrix-org/matrix-federation-tester/main.go [hunk 1]
```diff
@@ -129,14 +130,15 @@ func Report(
...
+
 	report.DNSResult = *dnsResult
 	// Map of network address to report.
 	report.ConnectionReports = make(map[string]ConnectionReport)
 	// Map of network address to connection error.
 	report.ConnectionErrors = make(map[string]error)
 	now := time.Now()
 	for _, addr := range report.DNSResult.Addrs {
-		keys, connState, err := gomatrixserverlib.FetchKeysDirect(serverName, addr, sni)
+		keys, connState, err := gomatrixserverlib.FetchKeysDirect(serverHost, addr, sni)
```
[the remark this answers is not in the document]
Comment (alexte) [reply src/github.com/matrix-org/matrix-federation-tester/main.go]: https://matrix.org/docs/spec/server_server/r0.1.1.html 
2.1:
...The target server must present a valid certificate for <delegated_hostname>...
that's what my patch does afaict

**Comment by alexte** [comment]
The bug is still present in https://matrix.org/federationtester/api/report?server_name=mail.at

File: src/github.com/matrix-org/matrix-federation-tester/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply src/github.com/matrix-org/matrix-federation-tester/main.go]: actually I think you're right, this line is valid. I'm not sure about the rest of the PR, though.

File: src/github.com/matrix-org/matrix-federation-tester/main.go [hunk 2]
```diff
@@ -155,8 +157,14 @@ func Report(
 			}
 		}
 
+		serverHostName:=serverHost
 		if directCert != nil {
-			valid, _ := gomatrixserverlib.IsValidCertificate(serverName, directCert, intermediateCerts)
+			hostPart,_,err:=net.SplitHostPort(string(serverHost))
```
Comment (richvdh) [inline src/github.com/matrix-org/matrix-federation-tester/main.go]: this is redundant now, I think.

Comment (alexte) [reply src/github.com/matrix-org/matrix-federation-tester/main.go]: ok

**Comment by alexte** [comment]
I updated the patch. But take care this patch does not work without the new matrixlib that strips the port part in IsValidCertificate

**Comment by anoadragon453** [comment]
@richvdh Simply `gb update github.com/matrix-org/gomatrixserverlib`

**Pull request merged**